File: regions/system/library/typescript/server-common/src/service-error.ts
import { ErrorCode } from './error-code.js';
import type { ErrorDetail } from './error-detail.js';
import { ErrorResponse } from './error-response.js';

/** サービスエラーの種類を表す型 */
export type ServiceErrorType =
  | 'not_found'
  | 'bad_request'
  | 'unauthorized'
  | 'forbidden'
  | 'conflict'
  | 'unprocessable_entity'
  | 'too_many_requests'
  | 'internal'
  | 'service_unavailable';

/**
 * HTTP ステータスコードにマッピングされるサービスエラークラス。
 * Rust の ServiceError enum に対応する TypeScript 実装。
 */
export class ServiceError extends Error {
  /** エラーの種類 */
  readonly type: ServiceErrorType;
  /** 構造化されたエラーコード */
  readonly code: ErrorCode;
  /** フィールドレベルの詳細情報 */
  readonly details: ErrorDetail[];

  /**
   * ServiceError を生成する。
   */
  constructor(
    type: ServiceErrorType,
    code: ErrorCode,
    message: string,
    details: ErrorDetail[] = [],
  ) {
    super(message);
    this.name = 'ServiceError';
    this.type = type;
    this.code = code;
    this.details = details;
  }

  // --- ファクトリメソッド ---

  /**
   * 404 Not Found エラーを生成する。
   */
  static notFound(service: string, message: string): ServiceError {
    return new ServiceError('not_found', ErrorCode.notFound(service), message);
  }

  /**
   * 400 Bad Request エラーを生成する。
   */
  static badRequest(service: string, message: string): ServiceError {
    return new ServiceError(
      'bad_request',
      ErrorCode.validation(service),
      message,
    );
  }

  /**
   * 400 Bad Request エラーを詳細情報付きで生成する。
   */
  static badRequestWithDetails(
    service: string,
    message: string,
    details: ErrorDetail[],
  ): ServiceError {
    return new ServiceError(
      'bad_request',
      ErrorCode.validation(service),
      message,
      details,
    );
  }

  /**
   * 401 Unauthorized エラーを生成する。
   */
  static unauthorized(service: string, message: string): ServiceError {
    return new ServiceError(
      'unauthorized',
      ErrorCode.unauthorized(service),
      message,
    );
  }

  /**
   * 403 Forbidden エラーを生成する。
   */
  static forbidden(service: string, message: string): ServiceError {
    return new ServiceError(
      'forbidden',
      ErrorCode.forbidden(service),
      message,
    );
  }

  /**
   * 409 Conflict エラーを生成する。
   */
  static conflict(service: string, message: string): ServiceError {
    return new ServiceError(
      'conflict',
      ErrorCode.conflict(service),
      message,
    );
  }

  /**
   * 422 Unprocessable Entity エラーを生成する（ビジネスルール違反）。
   */
  static unprocessableEntity(
    service: string,
    message: string,
  ): ServiceError {
    return new ServiceError(
      'unprocessable_entity',
      ErrorCode.unprocessable(service),
      message,
    );
  }

  /**
   * 429 Too Many Requests エラーを生成する（レートリミット超過）。
   */
  static tooManyRequests(service: string, message: string): ServiceError {
    return new ServiceError(
      'too_many_requests',
      ErrorCode.rateExceeded(service),
      message,
    );
  }

  /**
   * 500 Internal Server Error を生成する。
   */
  static internal(service: string, message: string): ServiceError {
    return new ServiceError(
      'internal',
      ErrorCode.internal(service),
      message,
    );
  }

  /**
   * 503 Service Unavailable エラーを生成する。
   */
  static serviceUnavailable(
    service: string,
    message: string,
  ): ServiceError {
    return new ServiceError(
      'service_unavailable',
      ErrorCode.serviceUnavailable(service),
      message,
    );
  }

  /**
   * ErrorResponse に変換する。
   * 詳細情報がある場合は withDetails、ない場合は create を使用する。
   */
  toErrorResponse(): ErrorResponse {
    if (this.details.length > 0) {
      return ErrorResponse.withDetails(
        this.code.value,
        this.message,
        this.details,
      );
    }
    return ErrorResponse.create(this.code.value, this.message);
  }

  /**
   * エラー種類に対応する HTTP ステータスコードを返す。
   */
  statusCode(): number {
    switch (this.type) {
      case 'not_found':
        return 404;
      case 'bad_request':
        return 400;
      case 'unauthorized':
        return 401;
      case 'forbidden':
        return 403;
      case 'conflict':
        return 409;
      case 'unprocessable_entity':
        return 422;
      case 'too_many_requests':
        return 429;
      case 'internal':
        return 500;
      case 'service_unavailable':
        return 503;
    }
  }
}

File: regions/system/library/dart/server_common/lib/src/service_error.dart
import 'error_code.dart';
import 'error_detail.dart';
import 'error_response.dart';

/// サービスエラーの種別。
/// HTTP ステータスコードに対応する。
enum ServiceErrorType {
  /// 404 Not Found
  notFound,

  /// 400 Bad Request
  badRequest,

  /// 401 Unauthorized
  unauthorized,

  /// 403 Forbidden
  forbidden,

  /// 409 Conflict
  conflict,

  /// 422 Unprocessable Entity
  unprocessableEntity,

  /// 429 Too Many Requests
  tooManyRequests,

  /// 500 Internal Server Error
  internal,

  /// 503 Service Unavailable
  serviceUnavailable,
}

/// 高レベルのサービスエラー。
/// HTTP ステータスコードに対応する構造化エラーコードとメッセージを保持する。
class ServiceError implements Exception {
  /// エラー種別
  final ServiceErrorType type;

  /// 構造化エラーコード
  final ErrorCode code;

  /// エラーメッセージ
  final String message;

  /// フィールドレベルのエラー詳細（オプション）
  final List<ErrorDetail> details;

  const ServiceError({
    required this.type,
    required this.code,
    required this.message,
    this.details = const [],
  });

  /// 指定サービスの NotFound エラーを生成する。
  factory ServiceError.notFound(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.notFound,
      code: ErrorCode.notFound(service),
      message: message,
    );
  }

  /// 指定サービスの BadRequest エラーを生成する。
  factory ServiceError.badRequest(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.badRequest,
      code: ErrorCode.validation(service),
      message: message,
    );
  }

  /// 指定サービスの Unauthorized エラーを生成する。
  factory ServiceError.unauthorized(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.unauthorized,
      code: ErrorCode.unauthorized(service),
      message: message,
    );
  }

  /// 指定サービスの Forbidden エラーを生成する。
  factory ServiceError.forbidden(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.forbidden,
      code: ErrorCode.forbidden(service),
      message: message,
    );
  }

  /// 指定サービスの Conflict エラーを生成する。
  factory ServiceError.conflict(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.conflict,
      code: ErrorCode.conflict(service),
      message: message,
    );
  }

  /// 指定サービスの UnprocessableEntity エラーを生成する。
  factory ServiceError.unprocessableEntity(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.unprocessableEntity,
      code: ErrorCode.unprocessable(service),
      message: message,
    );
  }

  /// 指定サービスの TooManyRequests エラーを生成する。
  factory ServiceError.tooManyRequests(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.tooManyRequests,
      code: ErrorCode.rateExceeded(service),
      message: message,
    );
  }

  /// 指定サービスの Internal エラーを生成する。
  factory ServiceError.internal(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.internal,
      code: ErrorCode.internal(service),
      message: message,
    );
  }

  /// 指定サービスの ServiceUnavailable エラーを生成する。
  factory ServiceError.serviceUnavailable(String service, String message) {
    return ServiceError(
      type: ServiceErrorType.serviceUnavailable,
      code: ErrorCode.serviceUnavailable(service),
      message: message,
    );
  }

  /// ErrorResponse に変換する。
  ErrorResponse toErrorResponse() {
    if (details.isNotEmpty) {
      return ErrorResponse.withDetails(code, message, details);
    }
    return ErrorResponse(code, message);
  }

  @override
  String toString() => 'ServiceError(${code.value}): $message';
}
